### PR TITLE
Vertical indirection

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

This PR introduces vertical indirections. In Pseudocode:
```
for all levels k:
  for all cells c:
    out[c,k] = in[c,vert_nbh[k]]
    out[c,k] = in[c,vert_nbh[k]+1] //additional shift
```
More complicated constructs are disallowed, e.g. `out[c,k] = in[c,vert_nbh[k+1]]`, `out[c,k] = in[c,vert_nbh[vert_nbh_inner[k]]]` and `out[c,k] = in[c,vert_nbh1[k+1]+vert_nbh2[k+1]]` would all be illegal.

### Design

A `VerticalOffset` class was introduced. What was formerly known as `verticalOffset` (i.e. the integer to shift k) was renamed to `verticalShift` and encapsulated into that `VerticalOffset` class, along with a `FieldAccessExpr` known as `verticalIndirection`. The interface prohibits direct construction of the `verticalIndirection`, the underlying constructor is wrapped by the `VerticalOffset` s.t. only `FieldAccessExpr`s without Offsets can be constructed. 

Unfortunately, some limitations of the Visitor infrastructure drove some design decisions, e.g.
* the `verticalIndirection` is actually stored as `std::shared_ptr<Expr>` because `getChildren()` returns `ArrayRef<std::shared_ptr<Expr>>`, i.e. access to the original `shared_ptr` is required due to the `ArrayRef`
* getting a _mutable_ reference `std::shared_ptr<Expr>& getIndirectionFieldAsExpr();` is required since there is no proper `const` visitor. 
These observations highlight the importance of [this issue](https://github.com/MeteoSwiss-APN/dawn/issues/617)

The presence of the vertical Indirection has deep consequences with regards to the `Extents` and `Intervals`, since they can both be constructed or modified with the `Offsets`. The presence of a vertical indirections implies that an **extent or an interval can become undefined**. This undefined state is either propagated, e.g. merging an offset with a an undefined offset makes the offset undefined, or simply leads to assertions, e.g. computing the gap intervals of a vector of intervals with some intervals being undefined is illegal. It was carefully evaluated with operation should propagate and which operations should be prohibited, in parts by the requirements imposed by the passes. However, there may still be edge cases we didn't catch. 

## Testing

* The usual slew of de/serialization tests for the new `Offsets` as well as the undefined value of the `Extent` (not required for `Interval` since only `SIR::Interval` is serialized, which does not carry an undefined state)
* Tests ensuring the propagation / prohibiting of `Extent` / `Interval` ops in undefined state.
* `Dawn4Py` tests exercising the newly introduced vertical indirections, also in presence of field versioning etc.
* In a preliminary step, a utility called `testMutator` was introduced to mutate IIR in such a way that each read becomes indirected. This was used to get a good idea about the affected parts of the code, but is currently unused. It remains in the PR in the hopes it may be useful for possible future debugging.

Unfortunately, `dusk` is not ready for vertical indirections yet. It is expected that going end to end might uncover some currently missed edge cases

## Dependencies

This PR is independent, but relies on a [previous refactoring](https://github.com/MeteoSwiss-APN/dawn/pull/1022) to work. 

## Resolves / Enhances

Fixes https://github.com/MeteoSwiss-APN/dawn/issues/979
Fixes https://github.com/MeteoSwiss-APN/dawn/issues/980

### CODE REVIEW

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

File: dawn/test/unit-test/dawn/IIR/TestIIRSerializer.cpp [hunk 1]
```diff
@@ -427,7 +429,9 @@ TEST_F(IIRSerializerTest, IIRTests) {
   stmt->setID(22);
   iir::Accesses stmtAccesses;
   iir::Extents extents(ast::Offsets{ast::cartesian});
+  iir::Extents undefinedExtents(iir::HorizontalExtent{ast::cartesian}, iir::UndefinedExtent{});
   stmtAccesses.addReadExtent(42, extents);
+  stmtAccesses.addReadExtent(43, undefinedExtents);
```
Comment (Stagno) [inline dawn/test/unit-test/dawn/IIR/TestIIRSerializer.cpp:434]: Serialization/deserialization routines won't do checks on the validity of the IR, but this is invalid IIR (no field has been set 43 as access id). I suggest, instead, to add a separate example of `ExprStmt` which contains a vertical indirection access pattern, correctly crafted so that it doesn't invalidate the IIR.

**Review by Stagno — CHANGES_REQUESTED** [review]
I see that you didn't change `UnstructuredDimensionChecker.cpp`. What about accesses of this kind:
```
f1_cells, f2_cells : Field[Cell]
kidx : Field[Edge]
f2_cells = f1_cells[kidx[k]]
```
Is there any code forbidding them somewhere?

File: dawn/src/dawn/CodeGen/CodeGen.cpp [hunk 2]
```diff
@@ -498,7 +498,11 @@ std::string extentToString(iir::Extents const& extents, ast::GridType const& gri
       result += "false";
   }
   auto const& vExtents = extents.verticalExtent();
-  result += ", " + std::to_string(vExtents.minus()) + "," + std::to_string(vExtents.plus());
+  if(!vExtents.isUndefined()) {
+    result += ", " + std::to_string(vExtents.minus()) + "," + std::to_string(vExtents.plus());
+  } else {
+    result += ", UNDEFINED , UNDEFINED";
```
[the remark this answers is not in the document]
Comment (Stagno) [reply dawn/src/dawn/CodeGen/CodeGen.cpp:504]: (unconscious pun btw)

File: dawn/test/integration-test/dawn4py-tests/CMakeLists.txt [hunk 3]
```diff
@@ -52,3 +52,4 @@ add_python_example(NAME loop_copy_stencil)
 add_python_example(NAME generate_empty_stage)
 add_python_example(NAME generate_versioned_field)
 add_python_example(NAME unstructured_masked_fields)
+add_python_example(NAME vertical_indirection)
```
[the remark this answers is not in the document]
Comment (mroethlin) [reply dawn/test/integration-test/dawn4py-tests/CMakeLists.txt:55]: I'll do one better and introduce an unstructured integration test!

**Comment by mroethlin** [comment]
> I see that you didn't change `UnstructuredDimensionChecker.cpp`. What about accesses of this kind:
> 
> ```
> f1_cells, f2_cells : Field[Cell]
> kidx : Field[Edge]
> f2_cells = f1_cells[kidx[k]]
> ```
> 
> Is there any code forbidding them somewhere?

Excellent catch. This was added to `UnstructuredDimensionChecker` and in the process uncovered another bug!

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins
